### PR TITLE
[gui] Fix vulkan glfw image count

### LINKED ISSUE #3601: test_ggui fails on Vulkan
To repro:
Build taichi with Vulkan on, then `python tests/run_tests.py -a vulkan -vr2 -t1 -k "ggui"`. or `python -m pytest tests/python/test_ggui.py::test_geometry_2d -s`, yields the following error:

```
================================================= test session starts =================================================
platform linux -- Python 3.9.7, pytest-6.2.5, py-1.11.0, pluggy-1.0.0                                                  
rootdir: /home/qiao/Taichi/taichi-m                                                                                    
plugins: cov-3.0.0, rerunfailures-10.2, xdist-2.4.0, forked-1.3.0                                                      
collecting ... [Taichi] version 0.8.6, llvm 10.0.0, commit cbdf5dde, linux, python 3.9.7                               
collected 1 item                                                                                                       
                                                                                                                       
tests/python/test_ggui.py [Taichi] Starting on arch=vulkan                                                             
[W 11/23/21 17:52:24.022 133483] [vulkan_device.cpp:buffer@535] Overriding last binding
[W 11/23/21 17:52:24.080 133483] [vulkan_device.cpp:buffer@535] Overriding last binding                                
[W 11/23/21 17:52:24.101 133483] [vulkan_device.cpp:buffer@535] Overriding last binding                                
[W 11/23/21 17:52:24.154 133483] [vulkan_device.cpp:buffer@535] Overriding last binding                                
[W 11/23/21 17:52:24.175 133483] [vulkan_device.cpp:buffer@535] Overriding last binding                     
[W 11/23/21 17:52:24.228 133483] [vulkan_device.cpp:buffer@535] Overriding last binding
python: /home/qiao/Taichi/taichi-m/external/imgui/backends/imgui_impl_vulkan.cpp:1040: bool ImGui_ImplVulkan_Init(ImGui
_ImplVulkan_InitInfo *, VkRenderPass): Assertion `info->MinImageCount >= 2' failed.        
[E 11/23/21 17:52:24.230 133483] Received signal 6 (Aborted)                                                           
                                                           
                                                                                                                       
***********************************     
* Taichi Compiler Stack Traceback *                        
***********************************                        
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so(+0x5c1597) [0x7f0bf1a37597]                                
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so(+0x6c67ee) [0x7f0bf1b3c7ee]                                
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so: taichi::Logger::error(std::__cxx11::basic_string<char, std
::char_traits<char>, std::allocator<char> > const&, bool)                                                              
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so(+0xa61161) [0x7f0bf1ed7161]                                
/lib/x86_64-linux-gnu/libc.so.6(+0x46210) [0x7f0c270da210] 
/lib/x86_64-linux-gnu/libc.so.6: gsignal
/lib/x86_64-linux-gnu/libc.so.6: abort                                                                                 
/lib/x86_64-linux-gnu/libc.so.6(+0x25729) [0x7f0c270b9729]                                                             
/lib/x86_64-linux-gnu/libc.so.6(+0x36f36) [0x7f0c270caf36] 
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so(+0x2fa9ffc) [0x7f0bf441fffc]
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so: taichi::ui::vulkan::Gui::init_render_resources(VkRenderPas
s_T*)                                                      
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so: taichi::ui::vulkan::Renderer::draw_frame(taichi::ui::vulka
n::Gui*)                                                                                                               
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so: taichi::ui::vulkan::Window::draw_frame()                  
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so: taichi::ui::vulkan::Window::write_image(std::__cxx11::basi
c_string<char, std::char_traits<char>, std::allocator<char> > const&)                                      
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so: taichi::ui::PyWindow::write_image(std::__cxx11::basic_stri
ng<char, std::char_traits<char>, std::allocator<char> > const&)                                                        
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so(+0x4aa866) [0x7f0bf1920866]
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so(+0x4aa7bf) [0x7f0bf19207bf]   
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so(+0x4aa466) [0x7f0bf1920466]    
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so(+0x4aa346) [0x7f0bf1920346]
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so(+0x4aa245) [0x7f0bf1920245]                                
/home/qiao/Taichi/taichi-m/python/taichi/lib/taichi_core.so(+0x473163) [0x7f0bf18e9163]                                
python(+0x174714) [0x5611b64f7714]                                                                                     
python(_PyObject_MakeTpCall+0x37f) [0x5611b64c79ef]        

```

Nvidia-driver: 470.57.02
RTX 3080.

### PR DESCRIPTION
**Pull request by bobcao3**
Fixes #3601

### CODE REVIEW

**Comment by AmesingFlank** [comment]
This wont fix #3601.

test_ggui is running in headless mode, which doesn't have a swap chain. In headless mode, there is only one single image that I created manually.

**Comment by bobcao3** [comment]
Then you will need to make 2 and simulate a swapchain, i think.

File: taichi/backends/vulkan/vulkan_device.cpp [hunk 1]
```diff
@@ -2039,6 +2039,10 @@ void VulkanSurface::destroy_swap_chain() {
   vkDestroySwapchainKHR(device_->vk_device(), swapchain_, nullptr);
 }
 
+int VulkanSurface::get_image_count() {
+  return swapchain_images_.size();
```
Comment (AmesingFlank) [inline taichi/backends/vulkan/vulkan_device.cpp:2043]: In headless mode, `swapchain_images_` is empty, and there's only a single image (`screenshot_image_`)

**Comment by AmesingFlank** [comment]
> Then you will need to make 2 and simulate a swapchain, i think.

Yep.

**Comment by bobcao3** [comment]
> > Then you will need to make 2 and simulate a swapchain, i think.
> 
> Yep.

wait, if we don't have a swapchain, why are we initialing the GLFW backend of ImGUI

**Comment by AmesingFlank** [comment]
> > > Then you will need to make 2 and simulate a swapchain, i think.
> > 
> > 
> > Yep.
> 
> wait, if we don't have a swapchain, why are we initialing the GLFW backend of ImGUI

We aren't. GLFW related init methods are only called when `show_window` is true.

**Comment by bobcao3** [comment]
In #3601 it's an assertion error in GLFW

**Comment by AmesingFlank** [comment]
> In #3601 it's an assertion error in GLFW

huh? the log says
```
python: /home/qiao/Taichi/taichi-m/external/imgui/backends/imgui_impl_vulkan.cpp:1040: bool ImGui_ImplVulkan_Init(ImGui
_ImplVulkan_InitInfo *, VkRenderPass): Assertion `info->MinImageCount >= 2' failed.  
```

**Comment by bobcao3** [comment]
oops sorry. I will try to add another image

**Review by qiao-bo — APPROVED** [review]
Thanks a lot!

File: taichi/backends/vulkan/vulkan_device.cpp (hunk 1 above)
Comment (qiao-bo) [reply taichi/backends/vulkan/vulkan_device.cpp:2045]: @AmesingFlank , should i mark this as resolved and get the PR in?

**Pull request merged**